pkg/autoscaler/stats_scraper.go

```
@@ -0,0 +1,142 @@
+	// StatMessage to the given channel.
+	Scrape(statsCh chan<- *StatMessage)
+}
+
+// cacheDisabledClient is a http client with cache disabled. It is shared by
+// every goruntime for a revision scraper.
+var cacheDisabledClient = &http.Client{
+	Transport: &http.Transport{
+		// Do not use the cached connection
+		DisableKeepAlives: true,
+	},
+	Timeout: 3 * time.Second,
+}
+
+// ServiceScraper scrapes Revision metrics via a K8S service by sampling. Which
+// pod to be picked up to serve the request is decided by K8S.
+type ServiceScraper struct {
+	httpClient *http.Client
+	url        string
+	metricKey  string
+	logger     *zap.SugaredLogger
+}
+
+// NewServiceScraper creates a new StatsScraper for the Revision which
+// the given Metric is responsible for.
+func NewServiceScraper(metric *Metric, logger *zap.SugaredLogger) (*ServiceScraper, error) {
+	return newServiceScraperWithClient(metric, logger, cacheDisabledClient)
+}
+
+func newServiceScraperWithClient(metric *Metric, logger *zap.SugaredLogger, httpClient *http.Client) (*ServiceScraper, error) {
+	revName := metric.Labels[serving.RevisionLabelKey]
+	if revName == "" {
+		return nil, fmt.Errorf("no Revision label found for Metric %s", metric.Name)
+	}
+
+	serviceName := reconciler.GetServingK8SServiceNameForObj(revName)
+	return &ServiceScraper{
+		httpClient: httpClient,
+		url:        fmt.Sprintf("http://%s.%s:9090/metrics", serviceName, metric.Namespace),
+		metricKey:  NewMetricKey(metric.Namespace, metric.Name),
+		logger:     logger,
+	}, nil
+}
+
+// Scrape call the destination service then send it
+// to the given stats chanel
+func (s *ServiceScraper) Scrape(statsCh chan<- *StatMessage) {
+	stat, err := s.scrapeViaURL()
+	if err != nil {
+		s.logger.Errorw("Failed to get metrics", zap.Error(err))
+		return
+	}
+
+	s.sendStatMessage(*stat, statsCh)
+}
+
+func (s *ServiceScraper) scrapeViaURL() (*Stat, error) {
+	req, err := http.NewRequest("GET", s.url, nil)
```

please don't ignore the error.

---

Good catch.